eo-phi-normalizer/src/Language/EO/Phi/Rules/Yaml.hs

```diff
@@ -44,20 +51,52 @@ data RuleTest = RuleTest
   }
   deriving (Generic, FromJSON, Show)
 
-data Condition = IsNF {nf :: [MetaId]}
-  deriving (Generic, FromJSON, Show)
+data Subset = Subset
+  { attrs :: [Attribute]
+  , bindings :: [Binding]
+  }
+  deriving (Generic, Show, FromJSON)
+data Condition
+  = IsNF {nf :: [MetaId]}
+  | IsSubset {subset :: Subset}
+  | NotSubset {not_subset :: Subset}
+  deriving (Generic, Show)
+instance FromJSON Condition where
+  parseJSON = genericParseJSON defaultOptions{sumEncoding = UntaggedValue}
 
 parseRuleSetFromFile :: FilePath -> IO RuleSet
 parseRuleSetFromFile = Yaml.decodeFileThrow
 
 convertRule :: Rule -> Common.Rule
-convertRule Rule{..} _ctx obj =
+convertRule Rule{..} ctx obj =
   [ obj'
   | subst <- matchObject pattern obj
+  , all (\cond -> checkCond ctx cond subst obj) when
   , obj' <- [applySubst subst result]
   -- TODO: check that obj' does not have any metavariables
   ]
 
+-- | Given a condition, and a substition from object matching
+--   tells whether the condition matches the object
+checkCond :: Common.Context -> Condition -> Subst -> Bool
+checkCond ctx (IsNF metaIds) subst = all (Common.isNF ctx . applySubst subst . MetaObject) metaIds
+checkCond _ctx (IsSubset (Subset attrs bindings)) subst = any (`hasAttr` substitutedBindings) substitutedAttrs
+ where
+  substitutedBindings = concatMap (applySubstBinding subst) bindings
+  substitutedAttrs = map (applySubstAttr subst) attrs
+checkCond ctx (NotSubset s) subst = not $ checkCond ctx (IsSubset s) subst
+
+hasAttr :: Attribute -> [Binding] -> Bool
+hasAttr attr = any (isAttr attr)
+ where
+  isAttr :: Attribute -> Binding -> Bool
+  isAttr a (AlphaBinding a' _) = a == a'
+  isAttr a (EmptyBinding a') = a == a'
+  -- FIXME: Delta and Lambda are not parsed as Alpha attributes
+  isAttr (Label (LabelId "Δ")) (DeltaBinding _) = True
+  isAttr (Label (LabelId "λ")) (LambdaBinding _) = True
```

Comment:
What to do about these 2 attributes? Options I thought of:
1. Update `LabelId` token to support Greek symbols
2. Add them as constructors to `Attribute` (not preferred)

Reply:
Option 1 seems ok to me.